# Video Data EM Notebook

In [203]:
import sys
import py_entitymatching as em
import pandas as pd
import os
import time

###################################################################
#KEY!!!! VARIABLE PREVENTS OVERWRITING LABELED SAMPLED DATA
###################################################################
GENERATE_NEW_LABELED_DATA = False



In [139]:
# Display the versions
print('python version: ' + sys.version )
print('pandas version: ' + pd.__version__ )
print('magellan version: ' + em.__version__ )

python version: 2.7.14 |Anaconda, Inc.| (default, Dec  7 2017, 11:07:58) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]
pandas version: 0.22.0
magellan version: 0.3.0


Matching two tables typically consists of the following three steps:

** 1. Reading the input tables **

** 2. Blocking the input tables to get a candidate set **

** 3. Matching the tuple pairs in the candidate set **

# Read input tables

In [146]:
# Get the paths
path_A = os.getcwd() + '/DATA/imdb3.csv'
path_B = os.getcwd() + '/DATA/thenumbers3.csv'
print(path_A)

/Users/dcompgriff/PrivateDocs/College Docs/Graduate School/University of Wisconson/School Docs/Spring 2017/839 Data Analysis/Project Stage 3/DATA/imdb3.csv


In [147]:
# Load csv files as dataframes and set the key attribute in the dataframe
A = em.read_csv_metadata(path_A, key='id')
B = em.read_csv_metadata(path_B, key='id')
A.head()

id                     title  year       mpaa  runtime  \
0   0               Toy Story 2  1999          G   92 min   
1   1    The Outlaw Josey Wales  1976         PG  135 min   
2   2            Monsters, Inc.  2001          G   92 min   
3   3  In the Heat of the Night  1967  Not Rated  110 min   
4   4         Chungking Express  1994      PG-13  102 min   

                         genres        director  \
0  Animation, Adventure, Comedy   John Lasseter   
1                       Western  Clint Eastwood   
2  Animation, Adventure, Comedy     Pete Docter   
3         Crime, Drama, Mystery  Norman Jewison   
4         Crime, Drama, Romance    Kar-Wai Wong   

                                                                             stars  \
0           Ash Brannon,Lee Unkrich,Tom Hanks,Tim Allen,Joan Cusack,Kelsey Grammer   
1                       Clint Eastwood,Sondra Locke,Chief Dan George,Bill McKinney   
2  David Silverman,Lee Unkrich,Billy Crystal,John Goodman,Mary Gibbs,Steve Buscemi   
3                                Sidney Poitier,Rod Steiger,Warren Oates,Lee Grant   
4                     Brigitte Lin,Takeshi Kaneshiro,Tony Chiu-Wai Leung,Faye Wong   

      gross  
0  $245.85M  
1   $31.80M  
2  $289.92M  
3   $24.38M  
4    $0.60M

In [148]:
B.head()

id                          title  year mpaa      runtime genres  \
0   0                        Effects  2005   -1           -1     -1   
1   1  Ek Haseena Thi Ek Deewana Tha  2017   -1  105 minutes  Drama   
2   2                     Ekk Albela  2016   -1           -1  Drama   
3   3                The Daisy Chain  2009    R           -1     -1   
4   4    Dali & I: The Surreal Story  2011   -1           -1  Drama   

            director      stars     gross  
0       Dusty Nelson         -1        -1  
1     Suneel Darshan         -1  $149,491  
2  Shekhar Sartandel         -1    $1,907  
3                 -1         -1        -1  
4      Andrew Niccol  Al Pacino        -1

### Check for null entries

In [149]:
B.isnull().sum()

id          0
title       0
year        0
mpaa        0
runtime     0
genres      0
director    0
stars       0
gross       0
dtype: int64

In [150]:
A.isnull().sum()

id          0
title       0
year        0
mpaa        0
runtime     0
genres      0
director    0
stars       0
gross       0
dtype: int64

### Check the data size

In [151]:
print('Number of tuples in A: ' + str(len(A)))
print('Number of tuples in B: ' + str(len(B)))
print('Number of tuples in A X B (i.e the cartesian product): ' + str(len(A)*len(B)))

Number of tuples in A: 4291
Number of tuples in B: 31006
Number of tuples in A X B (i.e the cartesian product): 133046746


In [152]:
# Display the keys of the input tables
em.get_key(A), em.get_key(B)

('id', 'id')

In [153]:
# If the tables are large we can downsample the tables like this
A1, B1 = em.down_sample(A, B, 200, 1, show_progress=False)
len(A1), len(B1)

# But for the purposes of this notebook, we will use the entire table A and B

(140, 200)

# Block tables to get candidate set

Before we do the matching, we would like to remove the obviously non-matching tuple pairs from the input tables. This would reduce the number of tuple pairs considered for matching.
*py_entitymatching* provides four different blockers: (1) attribute equivalence, (2) overlap, (3) rule-based, and (4) black-box. The user can mix and match these blockers to form a blocking sequence applied to input tables.

For the matching problem at hand, we know that two restaurants with different city names will not match. So we decide the apply blocking over names:

In [154]:
# Blocking plan

# A, B -- overlap blocker [title] --------------------|---> candidate set

In [155]:
# Create attribute equivalence blocker
ab = em.OverlapBlocker()

# Block using city attribute
C1 = ab.block_tables(A, B, 'title', 'title', 
                    l_output_attrs=['title', 'year', 'mpaa', 'runtime', 'genres', 'director', 'stars', 'gross'], 
                    r_output_attrs=['title', 'year', 'mpaa', 'runtime', 'genres', 'director', 'stars', 'gross'],
                    overlap_size=1
                    )

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:01:41


In [156]:
len(C1)

10956134

## Debug blocker output

The number of tuple pairs considered for matching is reduced to 10,956,134 (from 133,046,746), but we would want to make sure that the blocker did not drop any potential matches. We could debug the blocker output in *py_entitymatching* as follows:

In [157]:
# Debug blocker output
startTime = time.time()
dbg = em.debug_blocker(C1, A, B, output_size=200, attr_corres=[('title','title'), ('year', 'year')])
endTime = time.time()
print("Total time: %.2f seconds."%(endTime-startTime))

Total time: 44.68 seconds.


In [158]:
# Display first few tuple pairs from the debug_blocker's output
dbg.head(50)

_id  ltable_id  rtable_id      ltable_title ltable_year  \
0     0       2965      17445  Advise & Consent        1962   
1     1       2965      17842  Advise & Consent        1962   
2     2       2965      17844  Advise & Consent        1962   
3     3       2965      19318  Advise & Consent        1962   
4     4       2965      19600  Advise & Consent        1962   
5     5       2965      19796  Advise & Consent        1962   
6     6       2965      20161  Advise & Consent        1962   
7     7       2965      23293  Advise & Consent        1962   
8     8       2965      23439  Advise & Consent        1962   
9     9       2965      25285  Advise & Consent        1962   
10   10       2965      26599  Advise & Consent        1962   
11   11       2965      26620  Advise & Consent        1962   
12   12       2965      26824  Advise & Consent        1962   
13   13       2965      28999  Advise & Consent        1962   
14   14       2965      29445  Advise & Consent        1962   
15   15       2965      29519  Advise & Consent        1962   
16   16       2965      30408  Advise & Consent        1962   
17   17       3038        236      Love & Mercy        2014   
18   18       2661      29762              Chef        2014   
19   19       2661      30538              Chef        2014   
20   20       2661      30870              Chef        2014   
21   21       2965        236  Advise & Consent        1962   
22   22       2965        782  Advise & Consent        1962   
23   23       2965       4163  Advise & Consent        1962   
24   24       2965       4262  Advise & Consent        1962   
25   25       2965       4263  Advise & Consent        1962   
26   26       2965       6232  Advise & Consent        1962   
27   27       2965       6337  Advise & Consent        1962   
28   28       2965       6586  Advise & Consent        1962   
29   29       2965      13033  Advise & Consent        1962   
30   30       2965      13036  Advise & Consent        1962   
31   31       2965      13040  Advise & Consent        1962   
32   32       2965      14788  Advise & Consent        1962   
33   33       2965      15471  Advise & Consent        1962   
34   34       2965      16789  Advise & Consent        1962   
35   35       2965      16873  Advise & Consent        1962   
36   36       2965      16878  Advise & Consent        1962   
37   37       3038      23293      Love & Mercy        2014   
38   38       3038      23439      Love & Mercy        2014   
39   39       3038      25285      Love & Mercy        2014   
40   40       3038      26599      Love & Mercy        2014   
41   41       3038      26620      Love & Mercy        2014   
42   42       3038      26824      Love & Mercy        2014   
43   43       3038      28999      Love & Mercy        2014   
44   44       3038      29445      Love & Mercy        2014   
45   45       3038      29519      Love & Mercy        2014   
46   46       3038      30408      Love & Mercy        2014   
47   47       3431        236  Oliver & Company        1988   
48   48       3431        782  Oliver & Company        1988   
49   49       3431       4163  Oliver & Company        1988   

            rtable_title  rtable_year  
0          Ginger & Rosa         2013  
1            Harry & Son         1984  
2        Harry & Snowman         2016  
3        Lambert & Stamp         2015  
4          David & Layla         2007  
5        Hansel & Gretel         2002  
6                  Q & A         1990  
7        Starsky & Hutch         2004  
8        Spring & Arnaud         2016  
9         Town & Country         2001  
10          Withnail & I         1987  
11        Rodeo & Juliet         2016  
12           Rock & Rule         1983  
13         Water & Power         2014  
14      Wallace & Gromit         1996  
15             Up & Down         2005  
16        Yossi & Jagger         2003  
17          Burke & Hare         2011  
18              Ragnarok      

From the debug blocker's output we observe that the current blocker drops quite a few potential matches. We would want to update the blocking sequence to avoid dropping these potential matches.

For the considered dataset, we know that for the restaurants to match the  names must overlap between them. We could use overlap blocker for this purpose. Finally, we would want to union the outputs from the attribute equivalence blocker and the overlap blocker to get a consolidated candidate set.

In [159]:
# Updated blocking sequence
# A, B ------ overlap blocker [title] -----> C1--
#                                                     
# C1 ------ overlap blocker [year] --------> C2--

In [160]:
# Create overlap blocker
ob = em.OverlapBlocker()

# Block tables using 'name' attribute 
C2 = ob.block_candset(C1, 'year', 'year' 
                    )
len(C2)

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:03:37


134228

In [161]:
# Display first two rows from C2
C2.head(2)

_id  ltable_id  rtable_id    ltable_title ltable_year ltable_mpaa  \
34    34         45          3    The Hangover        2009           R   
144  144       4286          3  The Loved Ones        2009           R   

    ltable_runtime     ltable_genres ltable_director  \
34         100 min            Comedy   Todd Phillips   
144         84 min  Horror, Thriller      Sean Byrne   

                                                    ltable_stars ltable_gross  \
34       Zach Galifianakis,Bradley Cooper,Justin Bartha,Ed Helms     $277.32M   
144  Xavier Samuel,Robin McLeavy,Victoria Thaine,Jessica McNamee           -1   

        rtable_title  rtable_year rtable_mpaa rtable_runtime rtable_genres  \
34   The Daisy Chain         2009           R             -1            -1   
144  The Daisy Chain         2009           R             -1            -1   

    rtable_director rtable_stars rtable_gross  
34               -1           -1           -1  
144              -1           -1           -1

In [162]:
# Updated blocking sequence
# A, B ------ overlap blocker [title] -----> C1--
#                                                     
# C1 ------ overlap blocker [year] --------> C2--
#
# C2 ------ overlap blocker [mpaa] --------> C3--

In [163]:
# Create overlap blocker
ob = em.OverlapBlocker()

# Block tables using 'name' attribute 
C3 = ob.block_candset(C2, 'mpaa', 'mpaa' 
                    )
len(C3)

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:04


27767

In [164]:
# Display first two rows from C3
C3.head(2)

_id  ltable_id  rtable_id    ltable_title ltable_year ltable_mpaa  \
34    34         45          3    The Hangover        2009           R   
144  144       4286          3  The Loved Ones        2009           R   

    ltable_runtime     ltable_genres ltable_director  \
34         100 min            Comedy   Todd Phillips   
144         84 min  Horror, Thriller      Sean Byrne   

                                                    ltable_stars ltable_gross  \
34       Zach Galifianakis,Bradley Cooper,Justin Bartha,Ed Helms     $277.32M   
144  Xavier Samuel,Robin McLeavy,Victoria Thaine,Jessica McNamee           -1   

        rtable_title  rtable_year rtable_mpaa rtable_runtime rtable_genres  \
34   The Daisy Chain         2009           R             -1            -1   
144  The Daisy Chain         2009           R             -1            -1   

    rtable_director rtable_stars rtable_gross  
34               -1           -1           -1  
144              -1           -1           -1

In [326]:
# Create blocker to block tuples that only match based on 'the'.
def overlap_ignoring_words(ltuple, rtuple):
    words = set(['the'])
    # Remove ignore words set from strings.
    lostring = ltuple['title'].lower().split()
    rostring = rtuple['title'].lower().split()
    l_tokens = []
    r_tokens = []
    for word in lostring:
        if word.strip() not in words:
            l_tokens.append(word)
    for word in rostring:
        if word.strip() not in words:
            r_tokens.append(word)
    l_tokens = set(l_tokens)
    r_tokens = set(r_tokens)
        
    # Compute overlap.
    #l_tokens = set(list(map(lambda item: item.strip(), lstring.split())))
    #r_tokens = set(list(map(lambda item: item.strip(), rstring.split())))
    intersection = l_tokens.intersection(r_tokens)
    if len(intersection) >= 1:
        return False
    else:
        return True
    
# Create and apply blocker.
bb = em.BlackBoxBlocker()
bb.set_black_box_function(overlap_ignoring_words)
C4 = bb.block_candset(C3)    
C4.head()
    

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:03


_id  ltable_id  rtable_id  \
4146  4146       1757         11   
6671  6671       1093         20   
6891  6891       1723         20   
7619  7619       3795         20   
7647  7647       3883         20   

                                                    ltable_title ltable_year  \
4146                                                      Hamlet        1996   
6671  The Assassination of Jesse James by the Coward Robert Ford        2007   
6891              Sweeney Todd: The Demon Barber of Fleet Street        2007   
7619                                            30 Days of Night        2007   
7647                                       In the Valley of Elah        2007   

     ltable_mpaa ltable_runtime            ltable_genres  ltable_director  \
4146       PG-13        242 min                    Drama  Kenneth Branagh   
6671           R        160 min  Biography, Crime, Drama   Andrew Dominik   
6891           R        116 min   Drama, Horror, Musical       Tim Burton   
7619           R        113 min         Horror, Thriller      David Slade   
7647           R        121 min    Crime, Drama, Mystery      Paul Haggis   

                                                      ltable_stars  \
4146      Kenneth Branagh,Julie Christie,Derek Jacobi,Kate Winslet   
6671        Brad Pitt,Casey Affleck,Sam Shepard,Mary-Louise Parker   
6891   Johnny Depp,Helena Bonham Carter,Alan Rickman,Timothy Spall   
7619          Josh Hartnett,Melissa George,Danny Huston,Ben Foster   
7647  Tommy Lee Jones,Charlize Theron,Jonathan Tucker,Jason Patric   

     ltable_gross       rtable_title  rtable_year rtable_mpaa rtable_runtime  \
4146       $4.41M             Hamlet         1996       PG-13             -1   
6671       $3.90M  Home of the Brave         2007           R    106 minutes   
6891      $52.90M  Home of the Brave         2007           R    106 minutes   
7619      $39.57M  Home of the Brave         2007           R    106 minutes   
7647       $6.78M  Home of the Brave         2007           R    106 minutes   

     rtable_genres rtable_director  \
4146         Drama              -1   
6671         Drama   Irwin Winkler   
6891         Drama   Irwin Winkler   
7619         Drama   Irwin Winkler   
7647         Drama   Irwin Winkler   

                                                               rtable_stars  \
4146                                                        Kenneth Branagh   
6671  Samuel L. Jackson,Jessica Biel,Brian Presley,Curtis "50 Cent" Jackson   
6891  Samuel L. Jackson,Jessica Biel,Brian Presley,Curtis "50 Cent" Jackson   
7619  Samuel L. Jackson,Jessica Biel,Brian Presley,Curtis "50 Cent" Jackson   
7647  Samuel L. Jackson,Jessica Biel,Brian Presley,Curtis "50 Cent" Jackson   

     rtable_gross  
4146   $7,129,670  
6671           -1  
6891           -1  
7619           -1  
7647           -1

In [327]:
len(C4)

6334

We observe that the current blocker sequence does not drop obvious potential matches, and we can proceed with the matching step now. A subtle point to note here is, debugging blocker output practically provides a stopping criteria for modifying the blocker sequence.


# Matching tuple pairs in the candidate set

In this step, we would want to match the tuple pairs in the candidate set. Specifically, we use learning-based method for matching purposes.
This typically involves the following five steps:
1. Sampling and labeling the candidate set
2. Splitting the labeled data into development and evaluation set
3. Selecting the best learning based matcher using the development set
4. Evaluating the selected matcher using the evaluation set

## Sampling and labeling the candidate set

First, we randomly sample 600 tuple pairs for labeling purposes.

In [328]:
# Sample  candidate set
S = em.sample_table(C4, 600)
S.head(50)

_id  ltable_id  rtable_id  \
6891      6891       1723         20   
35988    35988       2769         74   
36081    36081       2472         74   
52331    52331       3476        116   
76674    76674       2498        154   
76739    76739       3222        160   
103328  103328       2296        256   
108966  108966       2552        295   
112898  112898        682        314   
116862  116862       2486        344   
116970  116970       1186        348   
129179  129179       1368        403   
143991  143991        526        464   
148487  148487       2681        487   
189240  189240        475        587   
244961  244961       4017        712   
246755  246755        500        724   
247201  247201       1739        724   
269154  269154       2807        778   
269175  269175       2915        778   
314538  314538        612        929   
317768  317768       3426        940   
318194  318194       3814        946   
402279  402279       1049       1139   
405698  405698       4065       1141   
546687  546687       1549       1516   
557701  557701       2520       1546   
567534  567534       1750       1568   
574124  574124       1291       1580   
587879  587879        456       1626   
588211  588211       2216       1631   
624205  624205        320       1760   
635495  635495       2108       1785   
640541  640541       2465       1789   
650403  650403       3064       1834   
687840  687840       1263       1964   
701664  701664       1121       2007   
706925  706925       3410       2017   
714853  714853         86       2060   
731686  731686       1461       2122   
796327  796327        728       2340   
796343  796343        781       2340   
796423  796423        989       2340   
804745  804745         57       2378   
835408  835408       2376       2485   
876034  876034       3579       2600   
876035  876035       1647       2602   
895500  895500        378       2670   
923998  923998       1037       2773   
976777  976777       3049       2908   

                                           ltable_title ltable_year  \
6891     Sweeney Todd: The Demon Barber of Fleet Street        2007   
35988                        Wristcutters: A Love Story        2006   
36081                                  A Scanner Darkly        2006   
52331                                             Boy A        2007   
76674           Mystery Science Theater 3000: The Movie        1996   
76739                                   Mystery, Alaska        1999   
103328                                           Buried        2010   
108966                                         Bulworth        1998   
112898                            Bullets Over Broadway        1994   
116862                                            Bugsy        1991   
116970                                     A Bug's Life        1998   
129179                      We Need to Talk About Kevin        2011   
143991                              Requiem for a Dream        2000   
148487                             The Science of Sleep        2006   
189240                The Diving Bell and the Butterfly        2007   
244961                                     Multiplicity        1996   
246755                                Shaun of the Dead        2004   
247201                                 Dawn of the Dead        2004   
269154                              The Opposite of Sex        1998   
269175                           The Last Days of Disco        1998   
314538                        Hedwig and the Angry Inch        2001   
317768                             For Love of the Game        1999   
318194                                            Cyrus        2010   
402279                                   Boys Don't Cry        1999   
405698  Going Clear: Scientology & the Prison of Belief        2015   
546687                                       Life of Pi        2012   
557701         Bad Lieutenant: Port of Call New Orlea

In [272]:
if GENERATE_NEW_LABELED_DATA:
    # Label S interactively. 
    G = em.label_table(S, 'gold')
    #SAVE! TIS A BITCH TO LABEL S AGAIN, AND IT NEEDS TO BE LABELED WITHIN THE UI!
    em.to_csv_metadata(G, './DATA/G.csv')
else:
    G = em.read_csv_metadata('./DATA/G.csv', 
                         key='_id',
                         ltable=A, rtable=B, 
                         fk_ltable='ltable_id', fk_rtable='rtable_id')


In [273]:
#em.to_csv_metadata(G, './DATA/G.csv')
#G2 = em.read_csv_metadata('./G2.csv', 
#                         key='_id',
#                         ltable=A, rtable=B, 
#                         fk_ltable='ltable_id', fk_rtable='rtable_id')

#em.save_table(G, './G2.pkl', metadata_ext='.pklmeta')
#G2 = em.load_table('./G2.pkl', metadata_ext='.pklmeta')
#catalog = em.get_catalog()
#em.show_properties(G2)

## Splitting the labeled data into development and evaluation set

In this step, we split the labeled data into two sets: development (I) and evaluation (J). Specifically, the development set is used to come up with the best learning-based matcher and the evaluation set used to evaluate the selected matcher on unseen data.

In [282]:
# Split S into development set (I) and evaluation set (J)
IJ = em.split_train_test(G, train_proportion=0.7, random_state=5)
I = IJ['train']
J = IJ['test']
print(len(I[I['gold']==1]))
print(len(J[J['gold']==1]))

33
11


## Selecting the best learning-based matcher 

Selecting the best learning-based matcher typically involves the following steps:

1. Creating a set of learning-based matchers
2. Creating features
3. Converting the development set into feature vectors
4. Selecting the best learning-based matcher using k-fold cross validation

### Creating a set of learning-based matchers

In [283]:
# Create a set of ML-matchers
dt = em.DTMatcher(name='DecisionTree', random_state=0)
svm = em.SVMMatcher(name='SVM', random_state=0)
rf = em.RFMatcher(name='RF', random_state=0)
lg = em.LogRegMatcher(name='LogReg', random_state=0)
ln = em.LinRegMatcher(name='LinReg')
nb = em.NBMatcher(name='NaiveBayes')

### Creating features

Next, we need to create a set of features for the development set. *py_entitymatching* provides a way to automatically generate features based on the attributes in the input tables. For the purposes of this guide, we use the automatically generated features.

In [284]:
# Generate features
feature_table = em.get_features_for_matching(A, B, validate_inferred_attr_types=False)

In [285]:
# List the names of the features generated
feature_table['feature_name']

0                                 id_id_exm
1                                 id_id_anm
2                            id_id_lev_dist
3                             id_id_lev_sim
4               title_title_jac_qgm_3_qgm_3
5           title_title_cos_dlm_dc0_dlm_dc0
6           title_title_jac_dlm_dc0_dlm_dc0
7                           title_title_mel
8                      title_title_lev_dist
9                       title_title_lev_sim
10                          title_title_nmw
11                           title_title_sw
12                mpaa_mpaa_jac_qgm_3_qgm_3
13            mpaa_mpaa_cos_dlm_dc0_dlm_dc0
14            mpaa_mpaa_jac_dlm_dc0_dlm_dc0
15                            mpaa_mpaa_mel
16                       mpaa_mpaa_lev_dist
17                        mpaa_mpaa_lev_sim
18                            mpaa_mpaa_nmw
19                             mpaa_mpaa_sw
20          runtime_runtime_jac_qgm_3_qgm_3
21      runtime_runtime_cos_dlm_dc0_dlm_dc0
22      runtime_runtime_jac_dlm_

### Converting the development set to feature vectors

In [286]:
# Convert the I into a set of feature vectors using F
H = em.extract_feature_vecs(I, 
                            feature_table=feature_table, 
                            attrs_after='gold',
                            show_progress=False)

In [287]:
# Display first few rows
H.head(3)

_id  ltable_id  rtable_id  id_id_exm  id_id_anm  id_id_lev_dist  \
89   1527164       2790       4352          0   0.641085               4   
347  6405280       2914      17394          0   0.167529               4   
472  9006279        649      25385          0   0.025566               5   

     id_id_lev_sim  title_title_jac_qgm_3_qgm_3  \
89             0.0                     0.000000   
347            0.2                     0.071429   
472            0.0                     1.000000   

     title_title_cos_dlm_dc0_dlm_dc0  title_title_jac_dlm_dc0_dlm_dc0  ...   \
89                          0.000000                         0.000000  ...    
347                         0.218218                         0.111111  ...    
472                         1.000000                         1.000000  ...    

     director_director_lev_sim  director_director_nmw  director_director_sw  \
89                    0.250000                   -1.0                   3.0   
347                   0.153846                    1.0                   3.0   
472                   1.000000                   14.0                  14.0   

     gross_gross_lev_dist  gross_gross_lev_sim  gross_gross_jar  \
89                    5.0             0.166667         0.555556   
347                  10.0             0.166667         0.541667   
472                  10.0             0.166667         0.472222   

     gross_gross_jwn  gross_gross_exm  gross_gross_jac_qgm_3_qgm_3  gold  
89          0.555556                0                     0.000000     0  
347         0.633333                0                     0.090909     0  
472         0.525000                0                     0.043478     1  

[3 rows x 54 columns]

### Selecting the best matcher using cross-validation

Now, we select the best matcher using k-fold cross-validation. For the purposes of this guide, we use five fold cross validation and use 'precision' and 'recall' metric to select the best matcher.

In [288]:
# Select the best ML matcher using CV
result = em.select_matcher([dt, rf, svm, ln, lg, nb], table=H, 
        exclude_attrs=['_id', 'ltable_id', 'rtable_id', 'gold'],
        k=5,
        target_attr='gold', metric_to_select_matcher='f1', random_state=0)
result['cv_stats']

/anaconda2/lib/python2.7/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/anaconda2/lib/python2.7/site-packages/scipy/linalg/basic.py:1226: RuntimeWarning: internal gelsd driver lwork query error, required iwork dimension not returned. This is likely the result of LAPACK bug 0038, fixed in LAPACK 3.2.2 (released July 21, 2010). Falling back to 'gelss' driver.
  warnings.warn(mesg, RuntimeWarning)
/anaconda2/lib/python2.7/site-packages/scipy/linalg/basic.py:1226: RuntimeWarning: internal gelsd driver lwork query error, required iwork dimension not returned. This is likely the result of LAPACK bug 0038, fixed in LAPACK 3.2.2 (released July 21, 2010). Falling back to 'gelss' driver.
  warnings.warn(mesg, RuntimeWarning)
/anaconda2/lib/python2.7/site-packages/scipy/linalg/basic.py:1226: RuntimeWarning: internal gelsd driver lwork query erro

Matcher  Average precision  Average recall  Average f1
0  DecisionTree           0.910000        0.946667    0.920498
1            RF           0.977778        0.913333    0.937709
2           SVM           0.800000        0.158333    0.258730
3        LinReg           0.977778        0.873333    0.912941
4        LogReg           0.927273        0.875000    0.880702
5    NaiveBayes           0.790729        0.966667    0.858768

### Debugging matcher

In [289]:
#  Split feature vectors into train and test
UV = em.split_train_test(H, train_proportion=0.5)
U = UV['train']
V = UV['test']

Next, we debug the matcher using GUI. For the purposes of this guide, we use random forest matcher for debugging purposes.

In [290]:
# Debug decision tree using GUI
# em.vis_debug_rf(rf, U, V, 
#        exclude_attrs=['_id', 'ltable_id', 'rtable_id', 'gold'],
#        target_attr='gold')

##  Evaluating the matching output

Evaluating the matching outputs for the evaluation set typically involves the following four steps:
1. Converting the evaluation set to feature vectors
2. Training matcher using the feature vectors extracted from the development set
3. Predicting the evaluation set using the trained matcher
4. Evaluating the predicted matches

### Converting the evaluation set to  feature vectors

As before, we convert to the feature vectors (using the feature table and the evaluation set)

In [291]:
# Convert J into a set of feature vectors using feature table
L = em.extract_feature_vecs(J, feature_table=feature_table,
                            attrs_after='gold', show_progress=False)

### Training the selected matcher

Now, we train the matcher using all of the feature vectors from the development set. For the purposes of this guide we use random forest as the selected matcher.

In [292]:
# Train using feature vectors from I 
rf.fit(table=H, 
       exclude_attrs=['_id', 'ltable_id', 'rtable_id', 'gold'], 
       target_attr='gold')

### Predicting the matches

Next, we predict the matches for the evaluation set (using the feature vectors extracted from it).

In [293]:
# Predict on L 
predictions = rf.predict(table=L, exclude_attrs=['_id', 'ltable_id', 'rtable_id', 'gold'], 
              append=True, target_attr='predicted', inplace=False)

### Evaluating the predictions

Finally, we evaluate the accuracy of predicted outputs

In [294]:
# Evaluate the predictions
eval_result = em.eval_matches(predictions, 'gold', 'predicted')
em.print_eval_summary(eval_result)

Precision : 91.67% (11/12)
Recall : 100.0% (11/11)
F1 : 95.65%
False positives : 1 (out of 12 positive predictions)
False negatives : 0 (out of 168 negative predictions)
